internal/thread/thread.go

```diff
@@ -14,59 +14,78 @@
 
 package thread
 
-import (
-	"errors"
-)
+import "sync"
 
 // Thread defines threading behavior in Ebiten.
 type Thread interface {
-	Call(func() error) error
+	Call(func())
 	Loop()
+	Stop()
+}
+
+type funcdata struct {
+	f    func()
+	done chan struct{}
 }
 
 // OSThread represents an OS thread.
 type OSThread struct {
-	funcs   chan func() error
-	results chan error
+	funcs chan funcdata
+	done  chan struct{}
 }
 
 // NewOSThread creates a new thread.
 //
 // It is assumed that the OS thread is fixed by runtime.LockOSThread when NewOSThread is called.
 func NewOSThread() *OSThread {
 	return &OSThread{
-		funcs:   make(chan func() error),
-		results: make(chan error),
+		funcs: make(chan funcdata),
+		done:  make(chan struct{}),
 	}
 }
 
-// BreakLoop represents an termination of the loop.
-var BreakLoop = errors.New("break loop")
-
-// Loop starts the thread loop until a posted function returns BreakLoop.
+// Loop starts the thread loop until Stop is called.
 //
 // Loop must be called on the thread.
 func (t *OSThread) Loop() {
-	for f := range t.funcs {
-		err := f()
-		if err == BreakLoop {
-			t.results <- nil
+	for {
+		select {
+		case fn := <-t.funcs:
+			fn.f()
+			if fn.done != nil {
+				fn.done <- struct{}{}
```

Comment:
I prefer having one channel like `funcdone` at the thread rather than the function.